power_grid_model_c/power_grid_model/include/power_grid_model/math_solver/math_solver.hpp

```diff
@@ -38,39 +38,28 @@ class MathSolver {
 
     MathOutput<sym> run_power_flow(PowerFlowInput<sym> const& input, double err_tol, Idx max_iter,
                                    CalculationInfo& calculation_info, CalculationMethod calculation_method) {
-        // set method to always linear if there are all load_gen with const_y
+        // set method to always linear if all load_gens have const_y
         calculation_method = all_const_y_ ? CalculationMethod::linear : calculation_method;
-        if (calculation_method == CalculationMethod::newton_raphson) {
-            if (!newton_pf_solver_.has_value()) {
-                Timer timer(calculation_info, 2210, "Create math solver");
-                newton_pf_solver_.emplace(y_bus_, topo_ptr_);
-            }
-            return newton_pf_solver_.value().run_power_flow(y_bus_, input, err_tol, max_iter, calculation_info);
-        }
-        else if (calculation_method == CalculationMethod::linear) {
-            if (!linear_pf_solver_.has_value()) {
-                Timer timer(calculation_info, 2210, "Create math solver");
-                linear_pf_solver_.emplace(y_bus_, topo_ptr_);
-            }
-            return linear_pf_solver_.value().run_power_flow(y_bus_, input, calculation_info);
-        }
 
-        else if (calculation_method == CalculationMethod::iterative_current ||
-                 calculation_method == CalculationMethod::linear_current) {
-            if (!iterative_current_pf_solver_.has_value()) {
-                Timer timer(calculation_info, 2210, "Create math solver");
-                iterative_current_pf_solver_.emplace(y_bus_, topo_ptr_);
-            }
-            if (calculation_method == CalculationMethod::linear_current) {
-                err_tol = 1000;
-                max_iter = 2;
+        try {
+            switch (calculation_method) {
+                case CalculationMethod::newton_raphson:
+                    return run_power_flow_newton_raphson(input, err_tol, max_iter, calculation_info);
+                case CalculationMethod::linear:
+                    return run_power_flow_linear(input, err_tol, max_iter, calculation_info);
+                case CalculationMethod::linear_current:
+                    return run_power_flow_linear_current(input, err_tol, max_iter, calculation_info);
+                case CalculationMethod::iterative_current:
+                    return run_power_flow_iterative_current(input, err_tol, max_iter, calculation_info);
+                default:
+                    throw InvalidCalculationMethod{};
             }
-            return iterative_current_pf_solver_.value().run_power_flow(y_bus_, input, err_tol, max_iter,
-                                                                       calculation_info);
         }
-
-        else {
-            throw InvalidCalculationMethod{};
+        catch (const SparseMatrixError&) {
```

Comment:
Catch the error here is a bit wired. You want the error to propagate above even when `err_tol = inf`. Users can theoretically run a  single (not-batch) calculation with set specifically `err_tol = inf`. We only need to catch and silence the error in the cache topology step in the batch calculation. So we can catch and silent there.

Reply:
well, i was thinking about the case where a user actually provides `err_tol = inf` where they really don't care about the outcome, but i'll change it. it does clean up the code a bit